apps/dashboard/app/(app)/apis/[apiId]/keys/[keyAuthId]/[keyId]/role-toggle.tsx

```diff
@@ -17,53 +17,71 @@ export const RoleToggle: React.FC<Props> = ({ roleId, keyId, checked }) => {
 
   const [optimisticChecked, setOptimisticChecked] = useState(checked);
   const connect = trpc.rbac.connectRoleToKey.useMutation({
-    onMutate: () => {
-      setOptimisticChecked(true);
-      toast.loading("Adding Role");
-    },
-    onSuccess: () => {
-      toast.success("Role added", {
-        description: "Changes may take up to 60 seconds to take effect.",
-        cancel: {
-          label: "Undo",
-          onClick: () => {
-            disconnect.mutate({ roleId, keyId });
-          },
-        },
-      });
-    },
-    onError(err) {
-      console.error(err);
-      toast.error(err.message);
-    },
     onSettled: () => {
       router.refresh();
     },
   });
   const disconnect = trpc.rbac.disconnectRoleFromKey.useMutation({
-    onMutate: () => {
-      setOptimisticChecked(false);
-      toast.loading("Removing role");
-    },
-    onSuccess: () => {
-      toast.success("Role removed", {
-        description: "Changes may take up to 60 seconds to take effect.",
-        cancel: {
-          label: "Undo",
-          onClick: () => {
-            connect.mutate({ roleId, keyId });
-          },
-        },
-      });
-    },
-    onError(err) {
-      console.error(err);
-      toast.error(err.message);
-    },
     onSettled: () => {
       router.refresh();
     },
   });
+
+  const handleConnect = async ({ roleId, keyId }: { roleId: string; keyId: string }) => {
+    setOptimisticChecked(true);
+    toast.promise(connect.mutateAsync({ roleId, keyId }), {
+      loading: "Adding Role ...",
+      success: () => (
```

Comment:
why are we doing custom components here?

the toast already has a way to display an action
https://sonner.emilkowal.ski/toast#action